secretariat/utils/outline.py

```diff
@@ -108,28 +119,83 @@ def find_user_from_email(self, email):
 
     def create_new_group(self, group_name):
         response = requests.post(
-            url=f"{self.url}/groups.create",
+            url=f"{self.api_url}/groups.create",
             headers=self.headers,
             json={"name": group_name},
         )
         if response.status_code >= 500:
             raise RemoteServerError(response.status_code)
 
         if response.status_code == 400:
-            object = response.json()
+            response = response.json()
             raise GroupCreationFailed(
-                400, f"{object.get('error')} - {object.get('message')}"
+                400, f"{response.get('error')} - {response.get('message')}"
             )
 
-        object = response.json()
-        group_uuid = object["data"]["id"]
+        group_uuid = response.json()["data"]["id"]
         return group_uuid
 
+    def list_groups(self, offset=0, limit=25):
+        response = requests.post(
+            url=f"{self.api_url}/groups.list",
+            headers=self.headers,
+            json={
+                "offset": offset,
+                "limit": limit,
+                "sort": "createdAt",
+                "direction": "ASC",
+            },
+        )
+        if response.status_code >= 500:
+            raise RemoteServerError(response.status_code)
+        return response.json().get("data").get("groups")
+
+    def find_group_by_name(self, group_name):
+        offset = 0
+        step = 20
+        group_list = self.list_groups(offset, step)
+        matching_groups = [group for group in group_list if group["name"] == group_name]
+
+        while len(group_list) and not (len(matching_groups)):
+            offset += step
+            group_list = self.list_groups(offset, step)
+            matching_groups = [
+                group for group in group_list if group["name"] == group_name
+            ]
+
+        if len(matching_groups):
+            return matching_groups[0]
+
+    def _request_list_memberships(self, group_id, offset, limit):
+        response = requests.post(
+            url=f"{self.api_url}/groups.memberships",
+            headers=self.headers,
+            json={
+                "offset": offset,
+                "limit": limit,
+                "sort": "createdAt",
+                "direction": "ASC",
+                "id": str(group_id),
+            },
+        )
+        if response.status_code >= 500:
+            raise RemoteServerError(response.status_code)
+        return response.json().get("data").get("users")
+
+    def list_group_users(self, group_id):
+        offset = 0
+        limit = 25
+        users = self._request_list_memberships(group_id, offset, limit)
+        while len(users):
+            yield from users
```

Comment:
Je ne sais pas si vous avez déjà vu le mot-clé `yield` ? Ça donne un générateur, ça permet à l'appelant de la méthode d'itérer directement sur le résultat sans devoir charger (ici) l'ensemble des users dans la mémoire. J'ai eu une sorte de révélation mystique quand j'ai découvert les générateurs en PHP, et j'ai été ravie de voir que ça existait aussi en Python. ^^

Reply:
Moi je l'avais déjà vu et utilisé mais je savais pas que ça faisait ça 🙃